File: tests/Core.CSharp.Tests/SchemaEvolutionTests.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text.Json;
using Xunit;
using Zeta.Core.CSharp;

namespace Zeta.Core.CSharp.Tests;

public class SchemaEvolutionTests
{
    private static DynamicValue ParseVal(JsonElement el)
    {
        var t = el.GetProperty("t").GetString();
        switch (t)
        {
            case "null":
                return new DynamicValue.Null();
            case "bool":
                return new DynamicValue.Bool(el.GetProperty("v").GetBoolean());
            case "int":
                return new DynamicValue.Int(long.Parse(el.GetProperty("v").GetString()!, CultureInfo.InvariantCulture));
            case "float":
                return new DynamicValue.Float(double.Parse(el.GetProperty("v").GetString()!, CultureInfo.InvariantCulture));
            case "str":
                return new DynamicValue.String(el.GetProperty("v").GetString()!);
            case "arr":
                var items = el.GetProperty("v").EnumerateArray().Select(ParseVal).ToImmutableArray();
                return new DynamicValue.Array(items);
            case "obj":
                var pairs = el.GetProperty("v").EnumerateArray().Select(item =>
                {
                    var key = item[0].GetString()!;
                    var val = ParseVal(item[1]);
                    return new KeyValuePair<string, DynamicValue>(key, val);
                }).ToImmutableArray();
                return new DynamicValue.Object(pairs);
            default:
                throw new NotSupportedException($"Unknown type {t}");
        }
    }

    private static SchemaEvolution.Migration ParseOp(JsonElement el)
    {
        var op = el.GetProperty("op").GetString();
        switch (op)
        {
            case "add":
                return SchemaEvolution.AddFieldMigration(0, el.GetProperty("key").GetString()!, ParseVal(el.GetProperty("default")));
            case "rename":
                return SchemaEvolution.RenameFieldMigration(0, el.GetProperty("from").GetString()!, el.GetProperty("to").GetString()!);
            case "remove":
                return SchemaEvolution.RemoveFieldMigration(0, el.GetProperty("key").GetString()!, ParseVal(el.GetProperty("default")));
            case "remove_with_dump":
                return SchemaEvolution.RemoveFieldWithDumpMigration(0, el.GetProperty("key").GetString()!);
            default:
                throw new NotSupportedException($"Unknown op {op}");
        }
    }

    [Fact]
    public void ReplaysGoldenVectorsSchemaEvolution()
    {
        var path = Path.GetFullPath(Path.Combine(AppContext.BaseDirectory, "../../../../src/Core.TypeScript/dynamic-value/golden-vectors-schema-evolution.json"));
        if (!File.Exists(path))
        {
            path = Path.GetFullPath(Path.Combine(AppContext.BaseDirectory, "../../../../../src/Core.TypeScript/dynamic-value/golden-vectors-schema-evolution.json"));
        }
        var json = File.ReadAllText(path);
        using var doc = JsonDocument.Parse(json);
        var root = doc.RootElement;
        var vectors = root.GetProperty("vectors").EnumerateArray();

        foreach (var vec in vectors)
        {
            var name = vec.GetProperty("name").GetString();
            var input = ParseVal(vec.GetProperty("input"));
            var expectedUp = ParseVal(vec.GetProperty("expected_up"));
            var expectedDown = ParseVal(vec.GetProperty("expected_down"));
            var ops = vec.GetProperty("ops").EnumerateArray().Select(ParseOp).ToList();

            // Run Up migrations
            var val = input;
            foreach (var op in ops)
            {
                val = op.Up(val);
            }
            Assert.True(expectedUp.Equals(val), $"Vector {name}: Up migration output mismatch");

            // Run Down migrations
            var backVal = val;
            for (int i = ops.Count - 1; i >= 0; i--)
            {
                var op = ops[i];
                Assert.NotNull(op.Down);
                backVal = op.Down(backVal);
            }
            Assert.True(expectedDown.Equals(backVal), $"Vector {name}: Down migration output mismatch");
        }
    }
}

File: src/Core.Rust.DynamicValue/tests/schema_evolution_tests.rs
//! Schema evolution cross-language conformance test replaying the shared seed.

use serde_json::Value;
use std::fs;
use std::path::PathBuf;
use zeta_core_dynamic_value::DynamicValue;
use zeta_core_dynamic_value::schema_evolution::{
    Migration, add_field_migration, remove_field_migration, remove_field_with_dump_migration,
    rename_field_migration,
};

fn repo_root() -> PathBuf {
    let mut dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    loop {
        if dir.join("Zeta.sln").exists() {
            return dir;
        }
        assert!(dir.pop(), "could not locate repo root (Zeta.sln)");
    }
}

fn build_value(v: &Value) -> DynamicValue {
    match v["t"].as_str().expect("tag string") {
        "null" => DynamicValue::Null,
        "bool" => DynamicValue::Bool(v["v"].as_bool().expect("bool value")),
        "int" => DynamicValue::Int(
            v["v"]
                .as_str()
                .expect("int decimal string")
                .parse::<i64>()
                .expect("i64 parse"),
        ),
        "float" => DynamicValue::Float(
            v["v"]
                .as_str()
                .expect("float decimal string")
                .parse::<f64>()
                .expect("f64 parse"),
        ),
        "str" => DynamicValue::String(v["v"].as_str().expect("str value").to_string()),
        "arr" => DynamicValue::Array(
            v["v"]
                .as_array()
                .expect("arr value")
                .iter()
                .map(build_value)
                .collect(),
        ),
        "obj" => DynamicValue::Object(
            v["v"]
                .as_array()
                .expect("obj value")
                .iter()
                .map(|pair| {
                    let p = pair.as_array().expect("pair array");
                    (
                        p[0].as_str().expect("key string").to_string(),
                        build_value(&p[1]),
                    )
                })
                .collect(),
        ),
        other => panic!("unknown tag {}", other),
    }
}

fn build_op(v: &Value) -> Migration {
    let op = v["op"].as_str().expect("op name");
    match op {
        "add" => {
            let key = v["key"].as_str().expect("key").to_string();
            let def = build_value(&v["default"]);
            add_field_migration(0, key, def)
        }
        "rename" => {
            let from = v["from"].as_str().expect("from").to_string();
            let to = v["to"].as_str().expect("to").to_string();
            rename_field_migration(0, from, to)
        }
        "remove" => {
            let key = v["key"].as_str().expect("key").to_string();
            let def = build_value(&v["default"]);
            remove_field_migration(0, key, def)
        }
        "remove_with_dump" => {
            let key = v["key"].as_str().expect("key").to_string();
            remove_field_with_dump_migration(0, key)
        }
        _ => panic!("unknown op {}", op),
    }
}

#[test]
fn replay_golden_vectors_schema_evolution() {
    let path =
        repo_root().join("src/Core.TypeScript/dynamic-value/golden-vectors-schema-evolution.json");
    let content = fs::read_to_string(path).expect("read golden vectors file");
    let root: Value = serde_json::from_str(&content).expect("parse json");
    let vectors = root["vectors"].as_array().expect("vectors array");

    for vec in vectors {
        let name = vec["name"].as_str().unwrap();
        let input = build_value(&vec["input"]);
        let expected_up = build_value(&vec["expected_up"]);
        let expected_down = build_value(&vec["expected_down"]);
        let ops: Vec<Migration> = vec["ops"]
            .as_array()
            .unwrap()
            .iter()
            .map(build_op)
            .collect();

        // Run Up migrations
        let mut val = input;
        for op in &ops {
            val = (op.up)(val);
        }
        assert_eq!(
            val, expected_up,
            "Vector {}: Up migration output mismatch",
            name
        );

        // Run Down migrations
        let mut back_val = val;
        for op in ops.iter().rev() {
            assert!(
                op.down.is_some(),
                "Vector {}: down migration function missing",
                name
            );
            if let Some(down) = &op.down {
                back_val = down(back_val);
            }
        }
        assert_eq!(
            back_val, expected_down,
            "Vector {}: Down migration output mismatch",
            name
        );
    }
}
